### ISSUE TITLE
Interaction between FlashbackAbility and X Castingcost spells with targeting requirements based on value of X

### ISSUE THREAD

**Issue by L0gi** [magefree/mage#2183]
first observed: unintended interaction between snapcaster granting flashback to Repeal.

when you first try to cast repeal from your graveyard it demands you to have a target with cmc 0, if you have no permanent with cmc 0 on the battlefield you can not cast repeal. If there is a permanent with cmc 0 on the battlefield, you can target it with repeal and then going forward you can choose the actual value for x and cast the spell as intended. 

This seems to be an issues in the interaction between these two cards specifically.

I have tested the interaction of snapcaster with blue suns zenith and the flashback works as intended, so it does not seem to be a general problem concerning flashback of cards with X in their mana costs.

I have also tested the interaction of snapcastermage with aurelias fury which works fine so it seems that it is not a general problem with X spells that require targets. Although the difference between aurelia*s fury and repeal in that aurelias fury targeting does not depend on the value of X in the same way that repeal does. 

I have tested also granting these spells flashback through jace, vryn's prodigi, and they all work as intended. 

So it seems that there is some difference in the way jvp and snapcaster mage grant flashback, or some other bug specific to the interaction between snapcaster and repeal.

This issues is relevant in that it affects some builds of modern blue based tron decks, which although not too popular are fun and beloved brews

**Comment by L0gi** [comment]
I just checked how Jace, Telepath unbound and snapacster grant flashback to a card and it seems that for whatever reason Jace, Telepaht Unbound uses his own implementation of "flashback". I assume therefore that the actual problem lies in the implementation of flashback. Unfortunately i do not know off the top of my head of any cards that would grant flashback to other spells similar to Snappcaster mage, to test this.

**Comment by L0gi** [comment]
One potential problem could be in how the flashback costs are added when granting the flashback ability. Does card.getCosts() respect mana consts with X correctly?

**Comment by L0gi** [comment]
I am not exactly sure yet about all internal works of the flashback ability and how exactly mana costs are represented, but can it be that the problem stems from line 199 of mage.abilities.keywords.flashbackAbility ?

`spellAbility.getManaCostsToPay().addAll(source.getManaCosts());`
what is the difference between X.getManaCostsToPay() and .getManaCosts()?
why is getManaCostsToPay() set to .getManaCosts() and not .getManaCostsToPay()?

I may be of course wrong and misinterpreting here, as I have just skimmed over the source and didn't dive too deep into yet.

**Comment by L0gi** [comment]
After some more testing:

the same issue can be replicated by granting flashback to spellblast with no 0cmc spell on the stack.

From this I assume that there is some problem in how the flashback costs are assigned/calculated as both problematic spells have their targeting requirements tied to their casting costs. 

I do not assume that the issue lies deeper in the way X spells and their targeting requirements are handled in general, as casting theses spells from hand works just fine, therefore the problem should be somwhere in how flashback costs are assigned to a spell.

**Comment by drmDev** [comment]
It is worth noting that Jace, Telepath Unbound does not grant a card Flashback. If it did, it would have used that keyword specifically. "You may cast target instant or sorcery card from your graveyard this turn. If that card would be put into your graveyard this turn, exile it instead."

There's been quite a few issues with flashback spell CMC calculation. It'll get fixed for the current issue, but then break something else we haven't tested yet.

We really need to get more tests in on various scenarios involving Flashback. There is a Mage.Tests project here where you add your own JUnit tests, if you aren't aware. Getting tests in when you find a bug is extremely good practice - that way whenever code is changed on Flashback to resolve this one case, all the tests can be run again to confirm it did not break another scenario. I've been trying to do this for every bug open. I'll try to add some tests in for some of the scenarios you mentioned.

**Comment by L0gi** [comment]
Yes, you are right with Jace, TU not granting flashback, but as it is a very similar effect it would use a similar internal mechanic to create this effect. Apperently I was wrong ^^'. 

I haven't worked with JUnit tests yet, if you could recommend somewhere where i can read up on them and your test framework API and  I would be interested in getting involved with you in development. And do you have any guidlines on where a specific test belongs?

As far as I understand, you have implemented some api that can be used for creating these tests. Setting up a game state, executing it and comparing it to some assertions. Is there any documentation on this api?

**Comment by L0gi** [comment]
So I think I've got the basic hang of your test API looking through other tests and have written a small test for the given scenario, 

how do I contribute that now to the main project?

**Comment by drmDev** [comment]
If you already have your local environment setup forked from the main xmage branch, you should be able to just commit and push your change to your repository. Then in your Github repo, there will be an option for submit a Pull Request to the XMage master branch where it can be reviewed and merged in.

**Comment by drmDev** [comment]
Not sure how far you are in the development process. If you are just getting started, check out the Wiki for: 
https://github.com/magefree/mage/wiki/Developer-Getting-Started
and 
https://github.com/magefree/mage/wiki/Developer-Testing-Tools

**Comment by L0gi** [comment]
I'm not getting any confirmation mail to create an account on the XMage forum. Are those sent out manually?

**Comment by L0gi** [comment]
after diving a bit deeper into the code i believe the problem lies deeper with the way flashback is implemented, not only the way flashback costs are calculated

flashback is not treated as an alternate casting cost, but somehow as an ability that is activated multiple times, first to allow the spell to be cast and the a second time to actually cast the spell and allow for mana payments. 

As flashback is not treated as alternate casting costs for the spell I suspected that there would be more problems with flashing back spells with alternate casting costs, 

I tested this with granting flashback to force of will, and lo and behold xmage actually does allow to flash back force of will for its alternate casting costs, which should not be possible. 

I am not sure how to create a good test for this scenario unfortunately. 

Is someone currently working on the flashback implementation? What reason was there to implement it as a "spell ability" (I am not sure what that relfects rules wise) instead of an Alternative Costs. I see an open redesing issue from 2014, has something changed concerning flashback since then, and who was working on it? 
I'd be willing to lend a hand here, but would really appreciate if someone who is more knowledgeable about the innerworkings of the engine could provide me with some top down overview of how everythings works together inside.
